### PR TITLE
fix: encode spaces in image_link URLs before Django validation

### LINKED ISSUE #1868: Unable to save chants in /source/699262
It's because of the image links.

I ran into this issue while I was creating a management command to update all chants in published sources.

### PR DESCRIPTION
**Pull request by PouyaMohseni**
- Replace spaces in the link before the validator runs
- Add unit tests for image link functionality

--------
Update: The previous form-level NormalizedURLField only normalized values that passed through the public edit forms. Django admin, management commands, and any direct ORM write still hit URLValidator with raw spaces and
failed.

Introduce NormalizedURLField (models.URLField subclass) used by Source.image_link and BaseChant.image_link. It normalizes in both to_python (full_clean) and get_prep_value (save), so every write path benefits. Its formfield() returns NormalizedURLFormField so ModelForms and admin auto-generated forms normalize user input too.

The field deconstructs back to models.URLField, so no migration is needed and the DB column is unchanged.

--------

resolves #1868

### CODE REVIEW

File: django/cantusdb_project/main_app/forms.py [hunk 1]
```diff
@@ -56,6 +56,15 @@
 ]
 
 
+class NormalizedURLField(forms.URLField):
+    """URLField that encodes spaces before Django's URLValidator fires."""
+
+    def to_python(self, value: Any) -> str:
+        if value:
+            value = value.strip().replace(" ", "%20")
```
Comment (ahankinson) [inline django/cantusdb_project/main_app/forms.py]: Is there a reason why you are not using a URL encoding library?

https://docs.python.org/3/library/urllib.parse.html#urllib.parse.quote

Comment (PouyaMohseni) [reply django/cantusdb_project/main_app/forms.py]: Because it would double encode a valid link unless I had an extensive safe-character management

File: django/cantusdb_project/main_app/models/url_field.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+from typing import Any, Optional
+
+from django import forms
+from django.db import models
+
+
+def _normalize(value: Any) -> Any:
```
Comment (ahankinson) [inline django/cantusdb_project/main_app/models/url_field.py]: Could you explain what this is doing? It’s returning None if the value is a string, but then using string methods on the value and returning that. Maybe instead of using Any you can narrow the types down on this?

File: django/cantusdb_project/main_app/forms.py [hunk 3]
```diff
@@ -1023,7 +1024,7 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         initial = kwargs.get("initial")
         if initial:
             for folio in initial:
-                self.fields[folio] = forms.CharField(
+                self.fields[folio] = NormalizedURLFormField(
```
Comment (ahankinson) [inline django/cantusdb_project/main_app/forms.py:1027]: Do you still need the form field if it’s normalizing at the model level?

Comment (PouyaMohseni) [reply django/cantusdb_project/main_app/forms.py:1027]: I believe yes, `ImageLinkForm.save()` calls `.update()`, so the model field covers the DB write. But Django runs URLValidator at form-validation time, before `.save()` is ever called. With a plain URLField here, a URL with spaces fails validation and the form returns a 200 instead of saving

File: django/cantusdb_project/main_app/models/url_field.py (hunk 2 above)
Comment (PouyaMohseni) [reply django/cantusdb_project/main_app/models/url_field.py]: it returns the encoded string when the value is a str, and passes none otherwise.

I narrowed the type down.

File: django/cantusdb_project/main_app/tests/test_functions.py [hunk 4]
```diff
@@ -490,3 +484,64 @@ def test_get_ci_text_search(self) -> None:
...
+            "https://example.com/Folio%2092r.jpg",
+        )
+
+    def test_deconstruct_reports_as_plain_urlfield(self):
+        # Keeps makemigrations from generating a no-op migration.
+        _, path, _, _ = self.field.deconstruct()
+        self.assertEqual(path, "django.db.models.URLField")
+
+    def test_formfield_returns_normalizing_form_field(self):
+        self.assertIsInstance(self.field.formfield(), NormalizedURLFormField)
```
Comment (liampond) [inline django/cantusdb_project/main_app/tests/test_functions.py]: Claude has suggested that we add another test — why not? The current tests check the field, but none of them try to save a model.

```
class ImageLinkSpaceEncodingIntegrationTest(TestCase):
    """End-to-end coverage for issue #1868: saving a model whose image_link
    contains spaces must succeed and persist the URL with spaces encoded."""

    def test_chant_with_spaced_image_link_saves(self) -> None:
        # Mirrors the management-command path from #1868: a direct .save(),
        # which BaseModel.save() routes through full_clean()/URLValidator.
        chant = make_fake_chant()
        chant.image_link = "https://example.com/Folio 92r.jpg"
        chant.save()
        chant.refresh_from_db()
        self.assertEqual(chant.image_link, "https://example.com/Folio%2092r.jpg")

    def test_source_with_spaced_image_link_saves(self) -> None:
        source = make_fake_source()
        source.image_link = "https://example.com/image gallery.jpg"
        source.save()
        source.refresh_from_db()
        self.assertEqual(source.image_link, "https://example.com/image%20gallery.jpg")

    def test_save_does_not_raise_on_spaces(self) -> None:
        # Without the fix this raises ValidationError ("Enter a valid URL"),
        # which is exactly the failure reported in #1868.
        chant = make_fake_chant()
        chant.image_link = "https://example.com/a b c.jpg"
        try:
            chant.save()
        except ValidationError:
            self.fail("Saving a chant with spaces in image_link should not raise.")
```

**Review by liampond — APPROVED** [review]
